src/Storages/StorageMongoDB.cpp

```diff
@@ -55,257 +60,327 @@ StorageMongoDB::StorageMongoDB(
 
-Pipe StorageMongoDB::read(
-    const Names & column_names,
-    const StorageSnapshotPtr & storage_snapshot,
-    SelectQueryInfo & /*query_info*/,
-    ContextPtr /*context*/,
-    QueryProcessingStage::Enum /*processed_stage*/,
-    size_t max_block_size,
-    size_t /*num_streams*/)
-{
-    connectIfNotConnected();
+    if (!context->getSettingsRef().allow_experimental_analyzer)
+    {
+        if (throw_on_error)
+            throw Exception(ErrorCodes::NOT_IMPLEMENTED, "MongoDB storage does not support 'allow_experimental_analyzer = 0' setting");
+        return make_document();
+    }
 
-    storage_snapshot->check(column_names);
+    const auto & query_tree = query.query_tree->as<QueryNode &>();
 
-    Block sample_block;
-    for (const String & column_name : column_names)
+    if (throw_on_error)
     {
-        auto column_data = storage_snapshot->metadata->getColumns().getPhysical(column_name);
-        sample_block.insert({ column_data.type, column_data.name });
+        if (query_tree.hasHaving())
+            throw Exception(ErrorCodes::NOT_IMPLEMENTED, "HAVING section is not supported. You can disable this error with 'SET mongodb_throw_on_unsupported_query=0', but this may cause poor performance, and is highly not recommended");
+        if (query_tree.hasGroupBy())
+            throw Exception(ErrorCodes::NOT_IMPLEMENTED, "GROUP BY section is not supported. You can disable this error with 'SET mongodb_throw_on_unsupported_query=0', but this may cause poor performance, and is highly not recommended");
+        if (query_tree.hasWindow())
+            throw Exception(ErrorCodes::NOT_IMPLEMENTED, "WINDOW section is not supported. You can disable this error with 'SET mongodb_throw_on_unsupported_query=0', but this may cause poor performance, and is highly not recommended");
+        if (query_tree.hasPrewhere())
+            throw Exception(ErrorCodes::NOT_IMPLEMENTED, "PREWHERE section is not supported. You can disable this error with 'SET mongodb_throw_on_unsupported_query=0', but this may cause poor performance, and is highly not recommended");
+        if (query_tree.hasLimitBy())
+            throw Exception(ErrorCodes::NOT_IMPLEMENTED, "LIMIT BY section is not supported. You can disable this error with 'SET mongodb_throw_on_unsupported_query=0', but this may cause poor performance, and is highly not recommended");
+        if (query_tree.hasOffset())
+            throw Exception(ErrorCodes::NOT_IMPLEMENTED, "OFFSET section is not supported. You can disable this error with 'SET mongodb_throw_on_unsupported_query=0', but this may cause poor performance, and is highly not recommended");
     }
 
-    return Pipe(std::make_shared<MongoDBSource>(connection, database_name, collection_name, Poco::MongoDB::Document{}, sample_block, max_block_size));
-}
+    auto on_error = [&] (const auto * node)
+    {
+        /// Reset limit, because if we omit ORDER BY, it should not be applied
```

Comment:
why? we can use limit without order by. for example, to select random rows. Mongo allows this.

Reply:
`on_error` is called for `where` as well, so we reset limit